### PR TITLE
Add free() for raft node

### LINKED ISSUE #419: Optimize memory footprint of raft state machine
After introducing [memory tracing](https://github.com/tikv/rfcs/pull/59) in TiKV, we found that the memory footprint could be very large if there were too many raft state machine. Most of them came from the inflight vector. Supposing `max_inflight_messages` was set to 4096, and there were 3 peers, leader would need to allocate 3 * 4096 * 8 = 96K memory. If there were 100k regions, then the virt memory became 9.1 GiB. And most of the memory were idle and perhaps never visited again.

There are many possible optimizations for this situation. On the one hand, if we can build a global flow control for all raft state machines, then the memory usage can be reduced; on the other hand, if we can reduce the total number of regions, the impact of flow control can also be omitted. But for the short term, I think it's easy to allocate lazily for inflight vector. We can use a linked list to allocate block by block and free them when it's idle. Because most raft state machines should be idle when it comes to 100k, so there will be only one block for all inflight vector. If each block is allocated as 512, then the memory footprint can be reduced to 572MiB, which is reasonably small enough.

### PR DESCRIPTION
**Pull request by jayzhan211**
Close #419 and https://github.com/tikv/raft-rs/pull/448 .

Add two API 
1. free_resources(id: u64)
2. free_resources_for_all()

Signed-off-by: jayzhan <jayzhan211@gmail.com>

### CODE REVIEW

**Comment by accelsao** [comment]
how to add test? or not

File: src/tracker/inflights.rs [hunk 1]
```diff
@@ -122,6 +122,12 @@ impl Inflights {
         self.count = 0;
         self.start = 0;
     }
+
+    /// free buffer memory
+    #[inline]
+    pub fn shrink(&mut self) {
+        self.buffer.shrink_to_fit();
```
Comment (accelsao) [inline src/tracker/inflights.rs:129]: After shrink the capacity will change,  the size for determine full() will be effected,  is it fine?

File: src/raft.rs [hunk 2]
```diff
@@ -2678,4 +2678,27 @@ impl<T: Storage> Raft<T> {
     pub fn uncommitted_size(&self) -> usize {
         self.uncommitted_state.uncommitted_size
     }
+
+    /// Free resources for all the raft node
+    pub fn free_resources_for_all(&mut self) {
```
Comment (BusyJay) [inline src/raft.rs:2683]: Just this method is enough. Choosing target id is too verbose.

File: src/tracker/inflights.rs (hunk 1 above)
Comment (BusyJay) [reply src/tracker/inflights.rs:129]: It's wrong. It should be able to reallocate when new messages need to be sent.

**Comment by BusyJay** [comment]
Tests should be added to check if flow control still work as expected before and after calling shrink.

File: src/tracker/inflights.rs [hunk 3]
```diff
@@ -122,6 +127,23 @@ impl Inflights {
         self.count = 0;
         self.start = 0;
     }
+
+    /// free buffer memory by reassignment
+    #[inline]
+    pub fn free_mem(&mut self) {
+        let mut v = Vec::with_capacity(self.count);
+        for i in 0..self.count {
+            v.push(self.buffer[self.start + i])
```
Comment (BusyJay) [inline src/tracker/inflights.rs:136]: I don't get it. `v` doesn't seem to be used, and why append the element from buffer to `v`? On the other hand, `self.buffer` should be dropped to free memory.

Comment (accelsao) [reply src/tracker/inflights.rs:136]: I am trying to re-create an `buffer` with `length` & `capacity` equal to `count`, in order to decrease the allocated memory, 
and remain only from `self.start` to `self.start + self.count - 1`

> On the other hand, `self.buffer` should be dropped to free memory

If  `self.count` is zero, `self.buffer` will be `Vec` with capacity 0, I am not sure what is the implementation inside, but I guess there is no memory allocated?
If  `self.count` is positive, we should at least remain `self.count` sized Vec (?

Comment (accelsao) [reply src/tracker/inflights.rs:136]: `shrink` that mentioned before, I think it will do nothing if the `Vec` is full once

Comment (jayzhan211) [reply src/tracker/inflights.rs:136]: What is the issue left related to the memory problem? Is #10156?

Comment (BusyJay) [reply src/tracker/inflights.rs:136]: What is 10156?

Comment (jayzhan211) [reply src/tracker/inflights.rs:136]: Sorry, is this https://github.com/tikv/tikv/issues/10156, found it under https://github.com/tikv/rfcs/pull/59

Comment (BusyJay) [reply src/tracker/inflights.rs:136]: /cc @hicqu

Comment (hicqu) [reply src/tracker/inflights.rs:136]: The left issue is unchanged for raft-rs. Memory usage of `Inflight` is too large if there are lots of raft peers.

**Comment by hicqu** [comment]
How about lazily allocate memory instead of adding `free_*` functions? @jayzhan211

**Comment by hicqu** [comment]
How about this solution:
for an given `Inflight`,
* allocate memory only if necessary
* reallocate memory if more items are pushed
* free and malloc again after `free_to` is called, and shrink the buffer if necessary

Or this:
* Allocate link-list nodes on heap, every node carries 16 u64 numbers

Any suggestions @BusyJay

File: src/tracker/inflights.rs (hunk 3 above)
Comment (hicqu) [reply src/tracker/inflights.rs:136]: @BusyJay do you mean `free_mem` should only free empty inflights, but keep no-empty inflights unchanged?

Comment (hicqu) [reply src/tracker/inflights.rs:136]: I have discussed with Jay. He thinks the current implemenatation is OK. BTW how about using `vec.extend_from_slice` instead?

File: src/tracker/inflights.rs [hunk 4]
```diff
@@ -122,6 +127,23 @@ impl Inflights {
...
+        }
+        self.buffer = v;
+        self.start = 0;
+    }
+
+    /// reallocate buffer memory
+    #[inline]
+    pub fn reallocate(&mut self) {
```
Comment (BusyJay) [inline src/tracker/inflights.rs:144]: Progress should reallocate when necessary.

Comment (jayzhan211) [reply src/tracker/inflights.rs:144]: Currently, buffer will be reallocated when you call this function, what is other situation when we need reallocation?

**Comment by hicqu** [comment]
@jayzhan211 we can try this solution:
* `free_mem` **shouldn't** re-allocate memory. So for an un-empty `Inflights`, we can
  1. leave it unchanged, or
  2. Call `shrink_to_fit` for it
  
  Both of them are OK.
* `Inflights::reallocate` is not necessary. Memory can be allocated when `Inflights::add` is called.

@BusyJay please help to confirm whether it fits with your design or not.

**Comment by hicqu** [comment]
@jayzhan211 any updates?

**Comment by jayzhan211** [comment]
> @jayzhan211 any updates?

Does it fit @BusyJay 's design

**Comment by BusyJay** [comment]
> @jayzhan211 we can try this solution:
> 
> * `free_mem` **shouldn't** re-allocate memory. So for an un-empty `Inflights`, we can
>   
>   1. leave it unchanged, or
>   2. Call `shrink_to_fit` for it
>   
>   Both of them are OK.
> * `Inflights::reallocate` is not necessary. Memory can be allocated when `Inflights::add` is called.
> * Call `free_mem` when a peer become follower.
> 
> @BusyJay please help to confirm whether it fits with your design or not.

LGTM, thanks!

**Comment by jayzhan211** [comment]
I have difficult for adding test (not famailiar with APIs).
Maybe add it afterall.

File: src/tracker/inflights.rs [hunk 5]
```diff
@@ -122,6 +127,19 @@ impl Inflights {
...
+
+    /// Free unused memory
+    #[inline]
+    pub fn free_mem(&mut self) {
+        // remove unused buffer, from start + count to the end of buffer
+        // ignore first index to self.start - 1
+        let len = self.buffer.len();
+        for _ in self.start + self.count..len {
+            self.buffer.pop();
+        }
```
Comment (jayzhan211) [inline src/tracker/inflights.rs:139]: We initialize the buffer with fixed capacity, so we need to remove element for reducing allocated memory, any better ways?

File: harness/tests/integration_cases/test_raft_flow_control.rs [hunk 6]
```diff
@@ -175,3 +175,24 @@ fn test_msg_app_flow_control_recv_heartbeat() {
...
+//     // force the progress to be in replicate state
+//     r.mut_prs().get_mut(2).unwrap().become_replicate();
+//
+//     // send message from 2 to 1 for adding inflights
+//     for _ in 0..10 {
+//         r.step(new_message(1, 1, MessageType::MsgPropose, 1))
+//             .expect("");
+//         r.read_messages();
+//     }
+//
```
Comment (jayzhan211) [inline harness/tests/integration_cases/test_raft_flow_control.rs:195]: After this line, inflights 2 (follower) is filled with 10 element, inflights 1 (leader) is empty, so it seems that we need to transfer leader from 2 to 1, then transfer 1 to 2 again, so peer 2 will call `free_mem` ?

File: src/tracker/inflights.rs (hunk 5 above)
Comment (hicqu) [reply src/tracker/inflights.rs:139]: How about don't touch it if it's not empty? So the resource can be freed later, and no allocations are introduced.

**Comment by hicqu** [comment]
@jayzhan211 I have send a patch PR to your code. Could you take a look?

File: src/tracker/inflights.rs (hunk 5 above)
Comment (jayzhan211) [reply src/tracker/inflights.rs:139]: > How about don't touch it if it's not empty? 

I am not sure how frequently the buffer is empty. If it happens a lot, this approach is good enough.

**Review by hicqu — APPROVED** [review]

File: harness/tests/integration_cases/test_raft_flow_control.rs [hunk 7]
```diff
@@ -175,3 +175,44 @@ fn test_msg_app_flow_control_recv_heartbeat() {
...
+            assert!(pr.ins.buffer_is_allocated());
+        }
+    }
+
+    let mut resp = new_message(2, 1, MessageType::MsgAppendResponse, 0);
+    resp.index = r.raft_log.last_index();
+    r.step(resp).unwrap();
+
+    // Test `maybe_free_inflight_buffers` works as expected.
+    r.maybe_free_inflight_buffers();
```
Comment (BusyJay) [inline harness/tests/integration_cases/test_raft_flow_control.rs:208]: It seems the case can pass even the function does nothing.

File: src/tracker/inflights.rs [hunk 8]
```diff
@@ -35,6 +38,7 @@ impl Clone for Inflights {
             start: self.start,
             count: self.count,
             buffer,
+            cap: self.cap,
```
Comment (BusyJay) [inline src/tracker/inflights.rs:41]: L36 can be also removed.

Comment (jayzhan211) [reply src/tracker/inflights.rs:41]: Then, I think we can use derive(Clone) for inflight

Comment (BusyJay) [reply src/tracker/inflights.rs:41]: Now that capacity is maintained explicitly, so the statement is not valid anymore. `derive(Clone)` is a good idea!.

File: harness/tests/integration_cases/test_raft_flow_control.rs [hunk 9]
```diff
@@ -175,3 +175,103 @@ fn test_msg_app_flow_control_recv_heartbeat() {
...
+
+    for (&id, pr) in r.prs().iter() {
+        if id != 1 {
+            assert!(pr.ins.buffer_is_allocated());
+        }
+    }
+
+    /*
+    1: cap=0/start=0/count=0/buffer=[]
+    2: cap=256/start=0/count=1/buffer=[2]
```
Comment (hicqu) [inline harness/tests/integration_cases/test_raft_flow_control.rs:207]: count should be 1?

File: harness/tests/integration_cases/test_raft_flow_control.rs [hunk 10]
```diff
@@ -175,3 +175,103 @@ fn test_msg_app_flow_control_recv_heartbeat() {
...
+    3: cap=256/start=0/count=1/buffer=[2]
+    */
+
+    let mut resp = new_message(2, 1, MessageType::MsgAppendResponse, 0);
+    resp.index = r.raft_log.last_index();
+    r.step(resp).unwrap();
+
+    for (&id, pr) in r.prs().iter() {
```
Comment (hicqu) [inline harness/tests/integration_cases/test_raft_flow_control.rs:215]: The loop is not necessary if only tests `id == 2`.

**Comment by jayzhan211** [comment]
I will use buffer.capacity, not cap(), because I found that we initialize buffer with `cap`

File: harness/tests/integration_cases/test_raft_flow_control.rs (hunk 9 above)
Comment (jayzhan211) [reply harness/tests/integration_cases/test_raft_flow_control.rs:208]: I am not sure why `MsgPropose` sends `2` to others, but it did.

**Comment by jayzhan211** [comment]
@hicqu PTAL

**Review by hicqu — DISMISSED** [review]

**Comment by hicqu** [comment]
@jayzhan211 seems some commits break the DCO check. Could you rebase and squash the PR like this:
```sh
git checkout free -b backup-free
git branch -D free
git checkout master -b free
git merge backup-free --no-commits --no-logs --squash
git commit -s     # then cleanup squashed commit logs
```

**Comment by jayzhan211** [comment]
> @jayzhan211 seems some commits break the DCO check. Could you rebase and squash the PR like this:
> 
> ```shell
> git checkout free -b backup-free
> git branch -D free
> git checkout master -b free
> git merge backup-free --no-commits --no-logs --squash
> git commit -s     # then cleanup squashed commit logs
> ```

Done

File: src/tracker/inflights.rs [hunk 11]
```diff
@@ -121,6 +112,29 @@ impl Inflights {
     pub fn reset(&mut self) {
         self.count = 0;
         self.start = 0;
+        self.buffer = vec![];
+    }
+
+    /// Number of inflight messages.
+    #[inline]
+    pub fn count(&self) -> usize {
```
Comment (BusyJay) [inline src/tracker/inflights.rs:120]: How about making it doc hiddern as it's only used in tests?

**Review by BusyJay — COMMENTED** [review]
@hicqu you may need a method to monitor the memory correctly.

**Comment by hicqu** [comment]
> @hicqu you may need a method to monitor the memory correctly.

Do you mean we need `fn Raft::inflight_buffers_size`?

**Comment by jayzhan211** [comment]
> > you may need a method to monitor the memory correctly.
> 
> @BusyJay Do you mean we need `fn Raft::inflight_buffers_size`?

Add a function that return `buffer.capacity()`?

**Comment by BusyJay** [comment]
> @BusyJay Do you mean we need `fn Raft::inflight_buffers_size`?

Yes, something like that.

**Comment by jayzhan211** [comment]
I think the buffer size will be either `0` or `cap`

**Comment by hicqu** [comment]
@jayzhan211 we need to add a new method for `Raft`:
```rust
+
+    /// Get the inflight buffer size.
+    pub fn inflight_buffers_size(&self) -> usize {
+        let mut total_size = 0;
+        for (_, pr) in self.prs().iter() {
+            total_size += pr.ins.buffer.capacity();
+        }
+        total_size
+    }
```

to get the allocated buffer size for a raft peer.

File: src/tracker/inflights.rs (hunk 11 above)
Comment (jayzhan211) [reply src/tracker/inflights.rs:120]: I got an error: missing documentation for an associated function

**Review by hicqu — APPROVED** [review]

**Review by BusyJay — APPROVED** [review]
Thanks!

**Comment by jayzhan211** [comment]
Thanks! @hicqu @BusyJay

**Pull request merged**